python/plugins/processing/algs/qgis/Climb.py

```diff
@@ -0,0 +1,280 @@
+
+        # Create new fields for climb and descent
+        thefields.append(QgsField(self.CLIMBATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.DESCENTATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MINELEVATTRIBUTE, QVariant.Double))
+        thefields.append(QgsField(self.MAXELEVATTRIBUTE, QVariant.Double))
+
+        layerwithz = source
+
+        (sink, dest_id) = self.parameterAsSink(parameters,
+                                               self.OUTPUT,
+                                               context, thefields,
+                                               layerwithz.wkbType(),
+                                               source.sourceCrs())
+
+        # get features from source (with z values)
+        features = layerwithz.getFeatures()
+        totalclimb = 0
+        totaldescent = 0
+        minelevation = float('Infinity')
+        maxelevation = float('-Infinity')
+
+        no_z_nodes = []
+
+        for current, feature in enumerate(features):
+            if feedback.isCanceled():
+                break
+            climb = 0
+            descent = 0
+            minelev = float('Infinity')
+            maxelev = float('-Infinity')
+            # In case of multigeometries we need to do the parts
+            parts = feature.geometry().constParts()
+            partnumber = 0
+            for part in parts:
+                # Calculate the climb
+                first = True
+                zval = 0
+                for idx, v in enumerate(part.vertices()):
+                    zval = v.z()
+                    if math.isnan(zval):
+                        no_z_nodes.append(self.tr(
+                            'Feature: {feature_id}, part: {part_id}, point: {point_id}'.format(
+                                feature_id=feature.id(),
+                                part_id=partnumber,
+                                point_id=idx)
+                        )
+                        )
+                        continue
+                    if first:
+                        prevz = zval
+                        minelev = zval
+                        maxelev = zval
+                        first = False
+                    else:
+                        diff = zval - prevz
+                        if diff > 0:
+                            climb = climb + diff
+                        else:
+                            descent = descent - diff
+                        if minelev > zval:
+                            minelev = zval
+                        if maxelev < zval:
+                            maxelev = zval
+                    prevz = zval
+                totalclimb = totalclimb + climb
+                totaldescent = totaldescent + descent
+                partnumber += 1
+            # Set the attribute values
+            attrs = feature.attributes()
+            outattrs = []
+            attrindex = 0
+            for attr in attrs:
```

Comment:
This block would be much simplified if you adapt the earlier suggestion of using combineFields -- then you'd just append the new attributes directly to the end of the existing ones.